src/core/scope.c

```diff
@@ -353,35 +359,87 @@ static void scope_enter_signal(Scope *s, ScopeState state, ScopeResult f) {
 
-        if (unit_has_name(u, SPECIAL_INIT_SCOPE))
-                return -EPERM;
+        (void) unit_realize_cgroup(u);
+        (void) unit_reset_accounting(u);
 
-        if (s->state == SCOPE_FAILED)
-                return -EPERM;
+        r = scope_arm_timer(s, usec_add(now(CLOCK_MONOTONIC), 30*USEC_PER_SEC));
+        if (r < 0)
+                return r;
 
-        /* We can't fulfill this right now, please try again later */
-        if (IN_SET(s->state, SCOPE_STOP_SIGTERM, SCOPE_STOP_SIGKILL))
-                return -EAGAIN;
+        r = unit_fork_helper_process(u, "(sd-chown-cgroup)", &pid);
+        if (r < 0)
+                goto fail;
 
-        assert(s->state == SCOPE_DEAD);
+        if (r == 0) {
+                uid_t uid = UID_INVALID;
+                gid_t gid = GID_INVALID;
 
-        if (!u->transient && !MANAGER_IS_RELOADING(u->manager))
-                return -ENOENT;
+                if (!isempty(s->user)) {
+                        const char *user = s->user;
+
+                        r = get_user_creds(&user, &uid, &gid, NULL, NULL, 0);
+                        if (r < 0) {
+                                log_unit_error_errno(UNIT(s), r, "Failed to resolve user \"%s\": %m", user);
+                                _exit(EXIT_USER);
+                        }
+                }
+
+                if (!isempty(s->group)) {
+                        const char *group = s->group;
+
+                        r = get_group_creds(&group, &gid, 0);
+                        if (r < 0) {
+                                log_unit_error_errno(UNIT(s), r, "Failed to resolve group \"%s\": %m", group);
+                                _exit(EXIT_GROUP);
+                        }
+                }
+
+                if (uid_is_valid(uid) && unit_cgroup_delegate(u)) {
+                        r = cg_set_access(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, uid, gid);
+                        if (r < 0) {
+                                log_unit_error_errno(UNIT(s), r, "Failed to adjust control group access: %m");
+                                _exit(EXIT_CGROUP);
+                        }
+                }
+
+                _exit(EXIT_SUCCESS);
+        }
+
+        r = unit_watch_pid(UNIT(s), pid, true);
+        if (r < 0)
+                goto fail;
+
+        scope_set_state(s, SCOPE_CHOWN_CGROUP);
+
+        return 1;
+fail:
+        s->timer_event_source = sd_event_source_disable_unref(s->timer_event_source);
+        return r;
+}
+
+static int scope_enter_running(Scope *s) {
+        Unit *u = UNIT(s);
+        int r;
+
+        assert(s);
 
         (void) bus_scope_track_controller(s);
 
         r = unit_acquire_invocation_id(u);
         if (r < 0)
                 return r;
 
-        (void) unit_realize_cgroup(u);
-        (void) unit_reset_accounting(u);
+        if (!u->cgroup_realized) {
+                (void) unit_realize_cgroup(u);
+                (void) unit_reset_accounting(u);
+        }
```

Comment:
Why do you move `unit_realize_cgroup()` into `scope_enter_running`? (It'd be simpler to keep it in `scope_start()`, no?)
Is the process attach vs chown order important?